File: src/test/java/kitchenpos/application/OrderTableServiceChangeNumberOfGuestsTest.java
package kitchenpos.application;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.NoSuchElementException;
import java.util.UUID;
import kitchenpos.application.fakerepository.OrderFakeRepository;
import kitchenpos.application.fakerepository.OrderTableFakeRepository;
import kitchenpos.domain.OrderRepository;
import kitchenpos.domain.OrderTable;
import kitchenpos.domain.OrderTableRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

@Nested
@DisplayName("OrderTableService의 ChangeNumberOfGuests메소드 테스트")
public class OrderTableServiceChangeNumberOfGuestsTest extends OrderTestSetup {

    private static final String TEST_NAME = "tableName";
    private static final int TEST_NUMBER_OF_GUEST = 2;

    private OrderTableRepository orderTableRepository;
    private OrderRepository orderRepository;

    private OrderTableService sut;

    @BeforeEach
    void setUp() {

        orderTableRepository = new OrderTableFakeRepository();
        orderRepository = new OrderFakeRepository();

        sut = new OrderTableService(orderTableRepository, orderRepository);
    }

    @Nested
    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
    class 메소드_호출시_예외_발생_조건_검사 {

        @Test
        void 요청_손님_수가_음수면_예외를_발생시킨다() {
            final OrderTable orderTable = orderTableRepository.save(create());
            final OrderTable request = create(-1);

            assertThatThrownBy(() -> sut.changeNumberOfGuests(orderTable.getId(), request))
                .isExactlyInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void orderTable이_없으면_예외를_발생시킨다() {
            final OrderTable request = create();

            assertThatThrownBy(() -> sut.changeNumberOfGuests(UUID.randomUUID(), request))
                .isExactlyInstanceOf(NoSuchElementException.class);
        }

        @Test
        void orderTable이_점유된_상태가_아니면_예외를_발생시킨다() {
            final OrderTable orderTable = orderTableRepository.save(create(false));
            final OrderTable request = create();

            assertThatThrownBy(() -> sut.changeNumberOfGuests(orderTable.getId(), request))
                .isExactlyInstanceOf(IllegalStateException.class);
        }

        private OrderTable create(final int numberOfGuests) {
            return createOrderTableRequest(TEST_NAME, numberOfGuests, false);
        }

        private OrderTable create() {
            return createOrderTableRequest(TEST_NAME, TEST_NUMBER_OF_GUEST, true);
        }

        private OrderTable create(final boolean occupied) {
            return createOrderTableRequest(TEST_NAME, TEST_NUMBER_OF_GUEST, occupied);
        }
    }
}

File: src/test/java/kitchenpos/application/OrderTableServiceChangeNumberOfGuestsTest.kt
package kitchenpos.application

import io.kotest.assertions.throwables.shouldThrowAny
import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeTypeOf
import io.mockk.mockk
import kitchenpos.domain.OrderRepository
import kitchenpos.domain.OrderTable
import kitchenpos.domain.OrderTableRepository
import kitchenpos.testsupport.FakeOrderTableRepository
import kitchenpos.testsupport.OrderTableFixtures

class OrderTableServiceChangeNumberOfGuestsTest : ShouldSpec({
    lateinit var savedOrderTable: OrderTable
    lateinit var orderTableRepository: OrderTableRepository
    lateinit var orderRepository: OrderRepository
    lateinit var service: OrderTableService

    beforeTest {
        orderTableRepository = FakeOrderTableRepository()
        orderRepository = mockk()
        service = OrderTableService(
            orderTableRepository,
            orderRepository
        )

        savedOrderTable = orderTableRepository.save(
            OrderTableFixtures.createOrderTable(
                isOccupied = true
            )
        )
    }

    context("주문 테이블 손님 수 변경") {
        should("성공") {
            // given
            val orderTableId = savedOrderTable.id
            val numberOfGuestsBeforeChange = savedOrderTable.numberOfGuests
            val request = createRequest(
                numberOfGuests = 2
            )

            // when
            val orderTableAfterChange = service.changeNumberOfGuests(
                orderTableId,
                request
            )

            // then
            numberOfGuestsBeforeChange shouldBe 0
            orderTableAfterChange.numberOfGuests shouldBe 2
        }

        should("실패 - 손님 수가 0 미만인 경우") {
            // given
            val orderTableId = savedOrderTable.id
            val request = createRequest(
                numberOfGuests = -1
            )

            // when
            val exception = shouldThrowAny {
                service.changeNumberOfGuests(
                    orderTableId,
                    request
                )
            }

            // then
            exception.shouldBeTypeOf<IllegalArgumentException>()
        }

        should("실패 - 손님이 착석한 테이블이 아닌 경우") {
            // given
            savedOrderTable = orderTableRepository.save(
                OrderTableFixtures.createOrderTable(
                    isOccupied = false
                )
            )
            val orderTableId = savedOrderTable.id
            val request = createRequest(
                numberOfGuests = 2
            )

            // when
            val exception = shouldThrowAny {
                service.changeNumberOfGuests(
                    orderTableId,
                    request
                )
            }

            // then
            exception.shouldBeTypeOf<IllegalStateException>()
        }
    }
}) {
    companion object {
        private fun createRequest(
            numberOfGuests: Int
        ): OrderTable {
            return OrderTable().apply {
                this.numberOfGuests = numberOfGuests
            }
        }
    }
}
